event-sourcing/event-store-jpa/src/main/java/org/apache/james/eventsourcing/eventstore/jpa/JPAEventStore.java

```diff
@@ -0,0 +1,176 @@
+     */
+    @Inject
+    @PersistenceUnit(unitName = "James")
+    public void setEntityManagerFactory(EntityManagerFactory entityManagerFactory) {
+        this.entityManagerFactory = entityManagerFactory;
+    }
+
+    @Override
+    public Publisher<Void> appendAll(Iterable<Event> events) {
+        if (events.isEmpty()) {
+            return Mono.empty();
+        }
+        Preconditions.checkArgument(Event.belongsToSameAggregate(events));
+        AggregateId aggregateId = events.head().getAggregateId();
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        final EntityTransaction transaction = entityManager.getTransaction();
+        try {
+            transaction.begin();
+            events.foreach(e -> {
+                try {
+                    JPAEvent jpaEvent = new JPAEvent(aggregateId, e.eventId(), jsonEventSerializer.serialize(e));
+                    entityManager.persist(jpaEvent);
+                } catch (JsonProcessingException ex) {
+                    throw new RuntimeException(ex);
+                }
+                return e;
+            });
+            transaction.commit();
+        } catch (PersistenceException e) {
+            if (transaction.isActive()) {
+                transaction.rollback();
+            }
+            EventStoreFailedException esfe = new EventStoreFailedException("Unable to add events");
+            esfe.initCause(e);
+            throw esfe;
+        } finally {
+            EntityManagerUtils.safelyClose(entityManager);
+        }
+        return Mono.empty();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Publisher<History> getEventsOfAggregate(AggregateId aggregateId) {
+        Preconditions.checkNotNull(aggregateId);
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        try {
+            List<Event> events = (List<Event>) entityManager.createNamedQuery(SELECT_AGGREGATE_QUERY)
+                .setParameter("aggregateId", aggregateId.asAggregateKey())
+                .getResultStream()
+                .map(e -> {
+                    try {
+                        return jsonEventSerializer.deserialize(((JPAEvent) e).getEvent());
+                    } catch (IOException ex) {
+                        throw new RuntimeException(ex);
+                    }
+                })
+                .collect(Collectors.toList());
+            return Mono.fromSupplier(() -> History.of(events.toArray(new Event[0])));
+        } catch (PersistenceException e) {
+            throw new RuntimeException("Unable to get events", e);
+        } finally {
+            EntityManagerUtils.safelyClose(entityManager);
+        }
+    }
+
+    @Override
+    public Publisher<Void> remove(AggregateId aggregateId) {
+        EntityManager entityManager = entityManagerFactory.createEntityManager();
+        final EntityTransaction transaction = entityManager.getTransaction();
```

Comment:
Use transaction runner here too and wrap the whole operation onto a mono...

Reply:
done